### PR TITLE
VTOL mission item handling

### LINKED ISSUE #3388: MC: land in mission should first move to the landing position before decent

### LINKED ISSUE #3414: VTOL: Support transition command for missions
@AndreasAntener How hard would it be to finish that off? I think it has been pending for some time and shouldn't be a major hurdle.

### LINKED ISSUE #3501: Standard VTOL issues
This is a list of all the issues of the standard VTOL code (many issues will be also applicable to the other VTOL types). The new simulation #3500 will allow to find and fix many of those issues.

 - [ ] Doing a transition in mission: In the simulation the plane starts to pitch up alot during the transition because the pitch setpoint dictates this. I believe the problem is that the multicopter still wants to stay in it's initial position and therefore commands pitch up. This problem was solved in position control mode which works fine in simulation.

 - [x] A mission which consists only of a takeoff and a transition to fw will end up in the plane flying into the ground because it's altitude setpoint will be set to 0 (relative to ground).

### LINKED ISSUE #3581: VTOL: Transition command likely interpreted as 3D waypoint
This log shows that an invalid setpoint was published at the point of transition:
http://logs.uaventure.com/view/pry8kgGojZtXxnqGQPx2SU

I'm suspecting that this was the transition command being mishandled.

### LINKED ISSUE #3583: Navigator: Move action handling into read routine, never load actions as mission items
@tumbili @AndreasAntener @julianoes Summarising my findings here again:

We check for action items here and publish the command:
https://github.com/PX4/Firmware/blob/0102e47708ff44d478cde919638b2f65b91ac0a6/src/modules/navigator/mission_block.cpp#L113-L127

The problem is that when we get there, the payload data from these items has already been taken and published as 3D coordinate. These are the bogus altitude values you see in VTOL.

The command publication code should be moved into this routine instead:
https://github.com/PX4/Firmware/blob/6eac78d67529e20c3170092fd271c11c57c70f2f/src/modules/navigator/mission.cpp#L658-L759

In the routine the code should cycle through the mission, publish all commands at it sees them (delaying between commands for 1000 us so they executed) and then return with the following 3D position it finds.

This way we will not have the twitch with VTOL any more. If you do this properly, these two PRs can be closed (as they try to fix it on the wrong location):
https://github.com/PX4/Firmware/pull/3579
this one should be harmless, but its a huge diff, so remove it rather as well:
https://github.com/PX4/Firmware/pull/3582

Let me know if this is not crystal clear and you need architectural help.

### PR DESCRIPTION
**Pull request by AndreasAntener**
In general this will fix handling for 2 part mission items (takeoff, landing, transitions)
Fixes #3583
Fixes #3581
Fixes #3501
Fixes #3414

Should handle the following cases consistently:
 - [x] transition: yaw into direction of next wp then transition
 - [x] back transition: transition then move to wp
 - [x] back transition yaw issue
 - [x] takeoff: climb then move to wp
 - [x] land: move to wp then descend (fixes #3388)
 - [x] back transition: yaw north when flying south

### CODE REVIEW

**Comment by julianoes** [comment]
@AndreasAntener ready for review?

**Comment by AndreasAntener** [comment]
Only conceptual, see the comment here: https://github.com/PX4/Firmware/issues/3583#issuecomment-174783402

**Comment by AndreasAntener** [comment]
@sanderux This is now working in STIL. Please take extra care when testing, I haven't thoroughly tested it yet, so please to a SITL test first. Forward transition now aligns towards next waypoint:

screen shot 2016-01-28 at 08 57 03

http://logs.uaventure.com/view/iMh8rPQvEyiCWGfNUTtq2C

**Comment by sanderux** [comment]
@AndreasAntener mission_block.cpp:195 
PX4_WARN("yaw err: %.3f", (float)yaw_err);

gives me a compilation error (double to float). disabled it

**Comment by sanderux** [comment]
@AndreasAntener I have a commit ready that fixes an IO hog because the mission feasibility checker was constantly running. Now only running before entering a mission. If you give me access i can attach to your PR to save a lot of merging.

**Comment by AndreasAntener** [comment]
Nice! Is it independent from this functionality and my changes? If yes we should treat it as such, so it's traceable and documented on a PR.

**Comment by sanderux** [comment]
On further inspection it seems although it loops it's only executed when a new home position is set. There does seem to be an issue of accepting invalid missions. see https://github.com/PX4/Firmware/issues/3627

**Comment by AndreasAntener** [comment]
Quick update, here is the log from our last flight: http://logs.uaventure.com/view/4G7pGXVjfMSttAmqPTCDkP

We had some issues with loading the mission back from QGC, which messed up the transition waypoints (lost the "frame"). So this flight http://logs.uaventure.com/view/BCMa7QKdCMgakcGdKFUSSg is actually also good but there was probably something wrong with the forward transition waypoint.

One or two more checks with latest QGC would be good and I need to go through the rest of the command handling to make sure nothing got lost during the refactoring.

**Comment by julianoes** [comment]
I think the mission feasablity checker needs to run when a mission is uploaded, not just before it gets executed. Otherwise the user doesn't get the feedback in time.
Or, is this already the case?

**Comment by sanderux** [comment]
@julianoes I think both should happen in a different way. I could be loading a mission from home, 10km away from my intended takeoff point. The FC could warn about the location but it should accept the mission. When switching to mission mode it should fail 10km away from the takeoff point. So i think both answers are correct but the feasibility checker should consider it's frame of reference (store or execute)

@AndreasAntener I have a flight window saturday and/or sunday, I will update to the latest QGC. is there anything specific i could test for you?

**Comment by AndreasAntener** [comment]
@sanderux Nothing specific. Would be cool to see this branch working for you as well. Still need to find some setpoint irregularities, curious if they show up in one of your logs as well.

On the feasibility checker, I think it already works like that. Or somewhat at leas. I was able to load a mission indoors (no position lock) and then go out and fly it. Maybe if you already have a lock it doesn't accept it then. Could you check these cases and open an issue for it? And describe how it should work.

**Comment by sanderux** [comment]
Full autonomous mission flown.
Log: http://logs.uaventure.com/view/5e9jP3aSS7Y4UaUyfqaYR9
Video: https://drive.google.com/open?id=0B27AvDmVit4KRVlpYkFMNVpTLWM

Saturation occurred at back transition, likely because the next waypoint commanded decent and the angle placed the vehicle sideways on fair wind. I don't know why the rotations before land occurred.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -904,9 +904,13 @@ MulticopterPositionControl::cross_sphere_line(const math::Vector<3> &sphere_c, f
 
 void MulticopterPositionControl::control_auto(float dt)
 {
-	if (!_mode_auto) {
+	if (!_mode_auto || _vehicle_status.in_transition_mode || !_vehicle_status.is_rotary_wing) {
```
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Please add a comment for why this condition is how it is. A reader has no clue what's going on here.

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 2]
```diff
@@ -904,9 +904,13 @@ MulticopterPositionControl::cross_sphere_line(const math::Vector<3> &sphere_c, f
...
 		_mode_auto = true;
 		/* reset position setpoint on AUTO mode activation */
```
Comment (julianoes) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Also, this comment is out of sync because it seems it's not just about AUTO.

**Comment by julianoes** [comment]
I had a quick look through the changes, mostly looking at comments and trying to spot stupid mistakes (without success which is good :smile:)

File: src/modules/navigator/mission.cpp [hunk 3]
```diff
@@ -404,50 +410,28 @@ Mission::set_mission_items()
...
+			has_next_position_item = true;
 
-		} else {
-			takeoff_alt = fmaxf(takeoff_alt, _navigator->get_home_position()->alt + _param_takeoff_alt.get());
-		}
+			float takeoff_alt = calculate_takeoff_altitude(&_mission_item);
 
-		/* check if we already above takeoff altitude */
-		if (_navigator->get_global_position()->alt < takeoff_alt) {
 			mavlink_log_critical(_navigator->get_mavlink_fd(), "takeoff to %.1f meters above home", (double)(takeoff_alt - _navigator->get_home_position()->alt));
```
Comment (LorenzMeier) [inline src/modules/navigator/mission.cpp]: This is not a critical message.

File: src/modules/navigator/mission.cpp [hunk 4]
```diff
@@ -732,7 +874,7 @@ Mission::read_mission_item(bool onboard, bool is_current, struct mission_item_s
 				*mission_index_ptr = mission_item_tmp.do_jump_mission_index;
 
 			} else {
-				if (is_current) {
+				if (offset == 0) {
 					mavlink_log_critical(_navigator->get_mavlink_fd(),
```
Comment (LorenzMeier) [inline src/modules/navigator/mission.cpp]: This is not a critical message.

**Comment by LorenzMeier** [comment]
@AndreasAntener Could you give this another pass? When is the next opportunity to test-fly the final state for you and / or @sanderux? If not today I would prefer to merge as-is to avoid branch drift.

**Comment by sanderux** [comment]
Wind conditions will make further testing hard for me this week. I have noticed several irregularities that i can reproduce in sitl. I will create mission files for this and share them with analysis and confirmation of such behavior in actual flight

**Comment by AndreasAntener** [comment]
@julianoes thanks!

@LorenzMeier I cannot test today, but let me correct a few things, also pushing default VTOL acceptance radius to 3m here as it seems that's what hits people currently the most.

**Comment by sanderux** [comment]
@AndreasAntener 
With reference to https://github.com/mavlink/qgroundcontrol/issues/3390 perhaps we could add the info log of the current waypoint as shown here: https://github.com/sanderux/Firmware/commit/c0f0d883eb9f20202708c9ba82ed90cdedcff1ee#diff-9b5a77642a641b107c8f8a966de34acdR821

Or shall i put this in a separate PR after merging this to master?

**Comment by sanderux** [comment]
@AndreasAntener please review https://github.com/UAVenture/Firmware/pull/2
this implements a general failsafe for most corner cases i encountered.

**Comment by sanderux** [comment]
Corner case in which the plane continues flight forever.

SITL waypoint file:
[land-and-then-backtransition.txt](https://github.com/PX4/Firmware/files/113944/land-and-then-backtransition.txt)

The sequence of mission items may seem unlikely but i will try to motivate my concern.
A user will experience that the transition command is executed at the start of the next waypoint (it might even seem as if it is executed before this point). The user could therefor attempt what might seem like a shortcut (set land and then transition) to have the plane land in VTOL.

The current result of this is a plane that continues forward on it's heading until it runs out of power or in to something. 

My suggestion is to have the feasibility checker validate that there will always be a waypoint after a transition command.

**Comment by AndreasAntener** [comment]
I'm just tracking down the yawing issues on takeoff/landing and @sanderux FW fly-away, should have an update soon.

**Comment by AndreasAntener** [comment]
@sanderux didn't yet find the cause of your landing/transition issue. The fixed wing landing already kicks in there.

Fixed a bunch of yawing issues, should be good to merge after another flight test.

**Comment by sanderux** [comment]
@AndreasAntener I have created a new PR for the front transition timeout (https://github.com/UAVenture/Firmware/pull/3) where i have integrated your and @LorenzMeier 's comments. 

The weather is very unstable but i will try to get some test flights done soon.

As soon as things have been tested and merged with master i will start work on mission failsafe (RTH) on failed transitions and  i could have a look at MAV_CMD_NAV_VTOL_TAKEOFF / LAND support.

**Comment by sanderux** [comment]
I updated the task list for 2 more issues. test can be done by flying this mission in sitl: 
[mission south.txt](https://github.com/PX4/Firmware/files/115354/mission.south.txt)

**Comment by AndreasAntener** [comment]
I added the back transition fix. For corner cases regarding missions "invalid" missions we should open new issues.

@sanderux could you give this another SITL test with valid missions?

**Comment by sanderux** [comment]
Tested. SITL normal mission looks great, no more unneeded yawing

**Comment by CornerOfSkyline** [comment]
I think we should make the velocity commander smooth. When the vtol translate back to the mc, the velocity can not decelerate smaller than the mc max velocity in the several translate back seconds, there will be a huge velocity commander to control the vtol in the mc mode. It cause a big pitch commander and the vtol respond the pitch commander , it looks danger when the vtol translate back.

**Comment by LorenzMeier** [comment]
The question is what velocity you commanded - if you command position hold (pitch stick zero) and trigger the velocity you get what you ask for - deceleration with the maximum braking moment. I haven't seen so far anything in real flight or in videos that I would consider extreme.

**Comment by AndreasAntener** [comment]
Two flights with the current state:
http://logs.uaventure.com/view/VxcqqMHj6cNiuqXDjcA9C7
http://logs.uaventure.com/view/GiVxcx93DbU3W5jFgXZ6Qc

Apart from the last bit they went great, there was a waypoint missing on the mission at the end, the climbing I cannot reproduce in SITL

@CornerOfSkyline yes I see pitching as well. Maybe something changed on master, I have a feeling this was better a while back.

**Comment by CornerOfSkyline** [comment]
@LorenzMeier  I test the VTOL auto mission in the gazebo_sitl , the whole fly is in mission mode. There is no manual control.

**Comment by sanderux** [comment]
@andreasantener does your last commit fully replace
https://github.com/PX4/Firmware/pull/3660 ?

**Comment by AndreasAntener** [comment]
@sanderux I rebased on master so it is extending it ;) just made this par more generic so you don't have to change that part when supporting new commands

**Comment by AndreasAntener** [comment]
Rebased and tested in SITL, also MC and FW missions. Awaiting last flight test.
